src/Psalm/Internal/Analyzer/Statements/Expression/Fetch/ArrayFetchAnalyzer.php

```diff
@@ -2069,4 +2053,61 @@ private static function handleArrayAccessOnString(
             }
         }
     }
+
+    /**
+     * @param Atomic[] $offset_types
+     */
+    private static function checkArrayOffsetType(
+        Type\Union $offset_type,
+        array $offset_types,
+        \Psalm\Codebase $codebase
+    ): bool {
+        $has_valid_absolute_offset = false;
+        foreach ($offset_types as $atomic_offset_type) {
+            if ($atomic_offset_type instanceof Type\Atomic\TClassConstant) {
+                $expanded = TypeExpander::expandAtomic(
+                    $codebase,
+                    $atomic_offset_type,
+                    $atomic_offset_type->fq_classlike_name,
+                    $atomic_offset_type->fq_classlike_name,
+                    null,
+                    true,
+                    true
+                );
+
+                if ($expanded instanceof Atomic) {
+                    $has_valid_absolute_offset = self::checkArrayOffsetType($offset_type, [$expanded], $codebase);
+                } else {
+                    $has_valid_absolute_offset = self::checkArrayOffsetType($offset_type, $expanded, $codebase);
+                }
+
+                if ($has_valid_absolute_offset) {
+                    break;
+                }
+            }
+
+            if ($atomic_offset_type instanceof Type\Atomic\TFalse &&
+                $offset_type->ignore_falsable_issues === true
+            ) {
+                //do nothing
+            } elseif ($atomic_offset_type instanceof Type\Atomic\TNull &&
+                $offset_type->ignore_nullable_issues === true
+            ) {
+                //do nothing
+            } elseif ($atomic_offset_type instanceof Type\Atomic\TString ||
+                $atomic_offset_type instanceof Type\Atomic\TInt ||
+                $atomic_offset_type instanceof Type\Atomic\TArrayKey ||
+                $atomic_offset_type instanceof Type\Atomic\TMixed ||
+                $atomic_offset_type instanceof Type\Atomic\TTemplateParam ||
+                (
+                    $atomic_offset_type instanceof Type\Atomic\TObjectWithProperties
+                    && isset($atomic_offset_type->methods['__toString'])
```

Comment:
What about `TNamedObject` with a `__toString()` method?

Reply:
fixed

Reply:
Well, turns out it's quite easy to add in check against array-key, but really hard against expected type.

It's really deep in the code and I didn't manage to make it work. If anyone wants to try, here's a reproducer https://psalm.dev/r/09a9137459

Reply:
oh well, turns out it's not even valid:
https://3v4l.org/PC88H

Reply:
Ah, I assumed it was valid because there was exiting code that allowed `object{__toString()}`. Array keys are really inconsistent in PHP :roll_eyes:

Reply:
Yeah, existing code was mine from a few weeks ago, I didn't check at the time...